src/app/ApplicationTemplate.Tests.Unit/AuthenticationServiceShould.cs

```diff
@@ -0,0 +1,189 @@
+﻿using System;
+using System.Net.Http;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
+using System.Threading;
+using System.Threading.Tasks;
+using ApplicationTemplate.Business;
+using ApplicationTemplate.DataAccess;
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+using Uno.Extensions;
+using Xunit;
+
+namespace ApplicationTemplate.Tests;
+
+/// <summary>
+/// Tests for the <see cref="AuthenticationService"/>.
+/// </summary>
+public sealed class AuthenticationServiceShould : IDisposable
+{
+	private readonly IAuthenticationService _authenticationService;
+	private readonly IApplicationSettingsRepository _applicationSettingsRepository;
+	private readonly IAuthenticationApiClient _authenticationRepository;
+	private readonly BehaviorSubject<ApplicationSettings> _applicationSettingsSubject;
+
+	public AuthenticationServiceShould()
+	{
+		_applicationSettingsRepository = Substitute.For<IApplicationSettingsRepository>();
+		_authenticationRepository = Substitute.For<IAuthenticationApiClient>();
+
+		_authenticationService = new AuthenticationService(
+			Substitute.For<ILoggerFactory>(),
+			_applicationSettingsRepository,
+			_authenticationRepository
+		);
+
+		_applicationSettingsSubject = new BehaviorSubject<ApplicationSettings>(CreateApplicationSettings());
+
+		_applicationSettingsRepository.GetAndObserveCurrent().Returns(_applicationSettingsSubject);
+		_applicationSettingsRepository.GetCurrent(CancellationToken.None).Returns(Task.FromResult(_applicationSettingsSubject.Value));
+
+		_applicationSettingsRepository.When(x => x.SetAuthenticationData(Arg.Any<CancellationToken>(), Arg.Any<AuthenticationData>()))
+			.Do(x => _applicationSettingsSubject.OnNext(CreateApplicationSettings(x.Arg<AuthenticationData>())));
+	}
+
+	[Fact]
+	public async Task ReturnAuthenticationData_WhenLoggingIn()
+	{
+		// Arrange
+		var email = "email";
+		var password = "password";
+		var authenticationData = CreateAuthenticationData(email: email);
+
+		_authenticationRepository.Login(Arg.Any<CancellationToken>(), Arg.Any<string>(), Arg.Any<string>()).Returns(authenticationData);
+
+		// Act
+		var result = await _authenticationService.Login(CancellationToken.None, email, password);
+
+		// Assert
+		Assert.Equal(authenticationData.AccessToken, result.AccessToken);
+	}
+
+	[Fact]
+	public async Task ReturnAuthenticationData_WhenRefreshingTheToken()
+	{
+		// Arrange
+		var email = "email";
+		var password = "password";
+		var refreshedEmail = "RefreshedEmail";
```

Comment:
refreshed emails?

Reply:
I don't think that should be how you test the refresh scenario because an email inside a token will likely never change. You could instead check the issued at and see it the new token is more recent.

Reply:
Yeah, I did not check them all; this is coming from one of our projects.